deepchem/feat/materials_featurizers.py

```diff
@@ -0,0 +1,285 @@
+    10.1103/PhysRevB.93.085142
+
+    Matminer: Ward, L. et al. Comput. Mater. Sci. 152, 60-69 (2018).
+
+    Pymatgen: Ong, S.P. et al. Comput. Mater. Sci. 68, 314-319 (2013). 
+
+  """
+
+  def __init__(self, data_source='matminer'):
+    """
+    Parameters
+    ----------
+    data_source : {"matminer", "magpie", "deml"}
+      Source for element property data.
+
+    """
+
+    self.data_source = data_source
+
+  def _featurize(self, comp):
+    """
+    Calculate chemical fingerprint from crystal composition.
+
+    Parameters
+    ----------
+    comp : str
+      Reduced formula of crystal.
+
+    Returns
+    -------
+    feats: np.ndarray
+      Vector of properties and statistics derived from chemical
+      stoichiometry.
+
+    """
+
+    from pymatgen import Composition
+    from matminer.featurizers.composition import ElementProperty
+
+    # Get pymatgen Composition object
+    c = Composition(comp)
+
+    ep = ElementProperty.from_preset(self.data_source)
+
+    try:
+      feats = ep.featurize(c)
+    except:
+      feats = []
+
+    return feats
+
+
+class SineCoulombMatrix(Featurizer):
+  """
+  Calculate sine Coulomb matrix for crystals.
+
+  A variant of Coulomb matrix for periodic crystals, based on 
+  Faber et al. Inter. J. Quantum Chem. 115, 16, (2015).
+
+  The sine Coulomb matrix is identical to the Coulomb matrix, except
+  that the inverse distance function is replaced by the inverse of
+  sin**2 of the vector between sites which are periodic in the 
+  dimensions of the crystal lattice.
+
+  Features are flattened into a vector of matrix eigenvalues by default
+  for ML-readiness. To ensure that all feature vectors are equal
+  length, the maximum number of atoms (eigenvalues) in the input
+  dataset must be specified.
+
+  This featurizer requires the optional dependency pymatgen. It may be
+  useful when crystal structures with 3D coordinates are available.
+
+  """
+
+  def __init__(self, max_atoms, eig=True):
+    """
+    Parameters
+    ----------
+    max_atoms : int
+      Maximum number of atoms for any crystal in the dataset. Used to
+      pad the Coulomb matrix.
+    eig : bool (default True)
+      Return flattened vector of matrix eigenvalues.
+
+    """
+
+    self.max_atoms = int(max_atoms)
+    self.eig = eig
+
+  def _featurize(self, struct):
+    """
+    Calculate sine Coulomb matrix from pymatgen structure.
+
+    Parameters
+    ----------
+    struct : dict
+      Json-serializable dictionary representation of pymatgen.core.structure
+      https://pymatgen.org/pymatgen.core.structure.html
+
+    Returns
+    -------
+    features: np.ndarray
+      2D sine Coulomb matrix, or 1D matrix eigenvalues. 
+
+    """
+
+    from pymatgen import Structure
+
+    s = Structure.from_dict(struct)
+    features = self.sine_coulomb_matrix(s)
+    features = np.asarray(features)
+
+    return features
+
+  def sine_coulomb_matrix(self, s):
```

Comment:
Here I didn't use matminer to limit the number of methods that require matminer as a soft dependency, and because the implementation of SCM is nearly identical to the one that is already in `dc.feat.coulomb_matrices`. I can certainly use the matminer module if that would be better.

Reply:
If matminer's codes have some mistakes and they are modified,  it is difficult to sync the modification.
So, i think it is better to use matminer's module directly.  

On the other hand, I think you know more deeply about this descriptor. If you seem this implementation is exactly correct, I think it is not necessary to use matminer's module directly. 

To be honest, I couldn't judge whether this implementation is definitely correct or not....  I confirmed the definition of SineCoulombMatrix from eq. 23 and 24 of [this thesis](https://arxiv.org/pdf/1503.07406.pdf), but I seem this implementation is a little tricky... (Especieally, I seem `trig_dist` calculation is a little strange... eq.24 includes the inverse matrix calculation, but such a caluculation is not seen in this implementation)

Reply:
Really good points in this thread. Doing some thinking, because of the similarity of `SCM` to `dc.feat.coulomb_matrices`, perhaps the right way to handle this is to expand the existing `CoulombMatrix` Featurizer to accept JSON crystal structures and use sine as the distance function.

@ncfrey Would it be straightforward to merge the SCM code into the existing Coulomb featurizer?

+1 for the idea of reusing the `matminer` implementation if possible if it works for our purposes. Less code to maintain makes life easy :)

Reply:
I think that if `Featurizer` and `CoulombMatrix` Featurizer are updated to accept Json files, then it should be no problem to generalize to SCM. For now, I'll move to the `matminer` implementation to keep things simple and easy to maintain.

Reply:
Makes sense! Perhaps we can revisit this in a future PR once the JSON file handling infrastructure is in place.